SE-2690 Python 3 Compatibility for Juniper Upgrade

nizarmah opened the pull request:
This PR updates XBlock Vimeo to work with python 3 for the juniper upgrade. In addition, it adds some information regarding installing and testing.

Tested on both Ironwood and Juniper.

**JIRA tickets**: SE-2690

**Dependencies**: None

**Screenshots**: 
There are no visual changes. But I made changes to the views and the rendering, so I am attaching screenshots in order to show that nothing wrong is happening.

![image](https://user-images.githubusercontent.com/5631091/83666539-c336ff80-a5d5-11ea-991b-a827ea3371b0.png)
![image](https://user-images.githubusercontent.com/5631091/83666598-d8139300-a5d5-11ea-9383-5c73b5989b34.png)
![image](https://user-images.githubusercontent.com/5631091/83666682-fd080600-a5d5-11ea-8e88-ea21edf63859.png)

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: "None"

**Testing instructions**:

### Manual Testing

#### Installing Vimeo XBlock

1. Create an *edX devstack* with branch `open-release/juniper.master`
2. Once the *edX devstack* is up, run `make studio-shell`
3. `source ../venvs/edxapp/bin/activate`
4. `sudo -u edxapp /edx/bin/pip.edxapp install "git+https://github.com/open-craft/xblock-vimeo@nizar/juniper_upgrade#egg=xblock-vimeo"`
5. `docker-compose restart studio`

#### Enabling Vimeo XBlock

1. In Studio, go to `Settings > Advanced Settings`.
1. Locate the `Advanced Module List` field, and add `"vimeo"` to that list.
1. Add a Unit to your course, and click on the `Advanced` button to add a `Vimeo` component.
1. Click `Edit` to add your Video URL, located on [vimeo.com](https://vimeo.com).
1. Click Save to add the video to your course.

--- 

### Automated Testing

1. Make sure you have `tox` installed.
2. Run the tox tests using the `tox` command.

**Reviewers**
- [ ] @pkulkark

nizarmah committed:
Switched UrlParse to UrlLib.Parse for Python 3

Added check for python version through `sys.version_info`. If it is python 3, then use urllib.parse instead of urlparse.

In python 3, the urlparse library was renamed to urllib.parse.

nizarmah committed:
Unified Fragment Creation in Method and Replaced Unicode with Resource String Method

Previously, each fragment was being created with 2 separate codes. This could be avoided by using a single method to create the fragment. In addition, the and use of unicode() isn't necessary. The string is already being utf8 decoded thanks to the resource_string method. Also, load_resource and resource_string were doing the same work, which was another reason to combine the fragment creation.

nizarmah committed:
Updated Setup to v0.5 and Added edX Release Tags

pkulkark commented on vimeo/vimeo.py:
```diff
@@ -53,6 +41,39 @@ def resource_string(self, path):
+        Returns:
+            Fragment 
+        """
+        html = Template(self.resource_string(html))
```
@nizarmah If python3 is not used, the html content will not be encoded with utf-8 by default. This could potentially break if the html content has unicode characters. Might make sense to add a check for the python version used and call `load_resource` if the version in below 3.

nizarmah committed:
Defined the Error Messages for Getting Embeded Code in Object

This will make it easier to change later on, without affecting the test cases, since the test cases will reference those error messagses from inside the object

nizarmah committed:
Added Tox Automated Testing with Unit Tests for Student and Studio View

nizarmah commented:
I added the Tox testing for `py{35}-django{22},py{38}-django{22,30}` without the `py38-quality`. I was having trouble with the `py38-quality`. I'll re-attempt if I end up having time on the ticket.

nizarmah committed:
Added Back load_resource for Python 2.7 HTML Content not Encoded with UTF-8

nizarmah committed:
Updated README with Testing Instructions and Installing XBlock on DevStack

nizarmah commented:
I had some trouble with the env `py{27}-django{18,111}` for`tox`. The issue was mainly because of the XBlock library. Some of the files in the library have classes created without referencing `object`. This was causing issues when I was trying to install the dependencies for python 2.7. 

The testing for 2.7 would work with no problem is that issue is resolved. The only thing that might have to happen is creating a new requirements file for python 2.7.

pkulkark commented on README.md:
```diff
@@ -19,7 +28,22 @@ Usage
+2. Once the *edX devstack* is up, run `make studio-shell`
+3. `source ../venvs/edxapp/bin/activate`
+4. `sudo -u edxapp /edx/bin/pip.edxapp install "git+https://github.com/open-craft/xblock-vimeo@nizar/juniper_upgrade#egg=xblock-vimeo"`
+5. `docker-compose restart studio`
```
Nit: The xblock will need to be installed in lms as well otherwise the unit will fail to load in lms.

nizarmah replied:
Oh, fair enough, I completely missed that. I tested the xblock on the Studio mainly without trying it on the LMS. I'll add a line for that.

pkulkark commented on tests/__init__.py:
```diff
@@ -0,0 +1,10 @@
+    field_data = DictFieldData({'data': 'Safe <b>html</b>'})
+    runtime = TestRuntime(field_data=field_data)
+
+    return VimeoBlock(runtime, scope_ids=None)
```
I would suggest moving this into a `utils.py` file instead of here.

pkulkark commented on tests/unit/test_xblock.py:
```diff
@@ -0,0 +1,70 @@
+        mock_embeded_code.return_value = ("", "")
+        mock_create_fragment.return_value = True
+
+        self.assertTrue(self.vimeo_block.student_view(None))
```
Please add a proper return_value to mocked method and verify them with the value returned in student_view so it's more clear what is being tested.

nizarmah replied:
I'll add an assert that it was called once instead of adding a return value. Since it should only be called once. What do you think @pkulkark?

nizarmah replied:
The `mock_embeded_code` is already being verified before. Therefore, there is no need to verify it because it already has unit tests. The `create_fragment` requires Django in order to render the things. Thus, I think creating an assert called once would be a good solution since it is called with a return statement.

pkulkark replied:
It's fine to add a return value since you're mocking the methods. The main idea is to verify that the student_view returns the correct value.

nizarmah replied:
I can't verify that student_view is returning the correct value because it requires that I render the template using `create_fragment`. And I can't test `create_fragment` without creating a runtime (the integration test). I already have an integration test setup for `create_fragment`. I just need to setup a runtime for it (but this will result in going overboard than the estimated time by like an hour).

But it is important to note that the `create_fragment` is a function which wouldn't be changed, it is kind of by the book. I am aware that it needs to be tested, however. But I can't test it without an instance.

> Actually I just had an idea, which might take 15 mins max. I will try to apply it. I'll mock django's rendering. And I'll verify that the fragment has the HTML code. I am not sure 100% it will work, but I'll give it a try.

pkulkark replied:
You don't need to test create_fragment here. The idea is to give a sense of what value we would see if this ran on an actual runtime. So instead of `self.assertTrue(self.vimeo_block.student_view(None))` you will assert it with what an actual result looks like (which is added to the return_value of the mocked create_fragment method). Basically it's for visualization. I hope that's clear?

nizarmah replied:
Yes, that's actually really clear. I'll work on it, thank you!

pkulkark commented on tests/unit/test_xblock.py:
```diff
@@ -0,0 +1,70 @@
+        mock_embeded_code.return_value = ("", "")
+        mock_create_fragment.return_value = True
+
+        self.assertTrue(self.vimeo_block.studio_view(context=None))
```
Same as above.

pkulkark commented on tests/unit/test_xblock.py:
```diff
@@ -0,0 +1,70 @@
+        self.assertTrue(self.vimeo_block.studio_view(context=None))
+
+if __name__ == "__main__":
+    unittest.main()
```
Nit: missing newline at the end

pkulkark commented on tox.ini:
```diff
@@ -0,0 +1,15 @@
+    django22: Django==2.2.12
+    django30: Django>=3.0,<3.1
+commands =
+    python -m unittest discover
```
Nit: missing newline at the end

nizarmah committed:
Added Extra Step to Run Installation on DevStack for LMS

Previously, it was only for Studio, but the installation should be done on both the LMS and the Studio

nizarmah committed:
Added Empty Line at the End of the Tox.ini File and Test File

nizarmah committed:
Moved XBlock Vimeo Test Instance Creator to Utils File

nizarmah committed:
Changed Mock Create Fragment from an Assert Return Value to an Assert Called Once Call for Studio and Student View

nizarmah committed:
Added Fragment Return Value to Mock Create Fragment

This makes it less confusing as to what the result should be at the end. The create_fragment method returns a fragment, therefore, its mock should too

pkulkark commented on tox.ini:
```diff
@@ -0,0 +1,15 @@
+
+[testenv]
+deps =
+	-r test_requirements.txt
```
Oh I also had to remove the space after `-r` otherwise I was getting an error ` No such file or directory: ' test_requirements.txt'`. So this needs to be changed to `-rtest_requirements.txt`

pkulkark reviewed (approved):
Thanks for sticking with this one @nizarmah. The changes look good to me. This is good to merge once the nit about the space is addressed.

- [x] I tested this: Verified the vimeo xblock works on juniper.master branch and tox tests run successfully.
- [x] I read through the code.
- [x] ~~I checked for accessibility issues~~ N/A
- [x] Includes documentation

nizarmah commented:
Do I squash and merge? Or is it the responsibility of the reviewer to merge the pull request? Please let me know @pkulkark

nizarmah committed:
Fixed Extra Space for Test Requirements File Dependency

pkulkark commented:
@nizarmah Yes please squash and merge. And be sure to add a proper final commit message. Also please remove the WIP tag before merging. You can merge once reviewer has approved the PR. If you don't have access to merge then let me know and I can do it for you.
